### DISCUSSION TITLE
Code completion

### DISCUSSION THREAD

**Discussion by cathaysia** [olimorris/codecompanion.nvim#352] [Ideas]
Will this plugin add code completion? Just like this: https://github.com/monkoose/neocodeium

**Comment by olimorris** [comment]
Probably not. It's something I have thought about greatly. But I doubt I'll be able to come close to what Copilot.vim or Copilot.lua do. Also, for anything other than ollama models, it's too expensive.

**Reply by cathaysia** [reply]
In my personal experience, Ollama code generation is slower than Copilot. What do you mean of expensive, do you mean that the requests to the server are too frequent?

**Reply by cathaysia** [reply]
If we increase the debounce time, may be able to reduce the cost?

**Reply by olimorris** [reply]
Expensive in terms of token cost. Have you found any models that can be connected to Ollama that do suggestions well?

My other reservation is that the LLM will need significant context about the workspace and the user's most recent edits, in order to determine the suggestion, which would take some time.

**Reply by cathaysia** [reply]
As of now, I haven't found a good code complete plugin used Ollama. The best chat plugin experience is this plugin.

Currently, codellama:7b is fast, with a delay of only about one second.

**Comment by dlants** [comment]
I think it's worth reconsidering. It could be disabled by default... It also could be hooked up to be manually prompted... like "finish what I'm trying to say here", but with a more immediate & polished UX than :cc ? Maybe there's already a way to hook codecompanion up to "finish what I'm trying to do here?

There's also an expanding set of things that this could lead to...

I think windsurf keeps track of all the edits in the file so far, and puts those in the context when sending to their model. They also seem to ask for edits in a wider window within the file, rather than just a continuation past the cursor, the result is a pretty slick experience that can often guess what you're intending to do next.

I'll try using this side-by-side with a manual copilot inline complete and report back :)

**Reply by olimorris** [reply]
If I change my mind on this I will let you all know. As I've stated before:

1) Copilot.lua is a fantastic plugin that likely won't be beaten by this. So I'm loathed to spend weeks implementing something which I won't use.

2) The workspace context needs to come first (I'm slowly working on it) and will need a lot of trial and error and continued output being sent to the LLM...

3) Which is expensive. Sure, users can turn it off but then I've created a feature that took weeks which very few users will leverage...

4) Because Ollama models are currently still way below par for most average laptop users.

**Reply by towry** [reply]
> Which is expensive. Sure, users can turn it off but then I've created a feature that took weeks which very few users will leverage...

I see many people on cursor forum asking for a neovim plugin and they would pay extra money for it, I have same feeling.

Codecompanion is great when I start to use it for specific task, but not great for continuous code writing, so I have this plugin Installed but barely use.


https://forum.cursor.com/t/cursor-plugin-for-neovim/6234

**Reply by olimorris** [reply]
Hopefully this will change when I release the workspace slash command.

**Comment by PLAZMAMA** [comment]
FYI, if you want code completion with local LLMs, checkout [Tabby](https://github.com/TabbyML/vim-tabby)

**Reply by Dronakurl** [reply]
Thanks for linking tabby. I like it. Do you know, if it can also be used as a chat agent in codecompanion?

**Reply by PLAZMAMA** [reply]
Not to my knowledge, but if you want to use a local LLM with code companion I would suggest using Ollama, which this plugin supports.

**Comment by Dronakurl** [comment]
Thanks, I did this already, but tabby has this support for context from own
repositories and documentation of frameworks that I want to use. It’s
possible to write own adapaters, so i should look into it

Maor ***@***.***> schrieb am Sa. 18. Jan. 2025 um 20:33:

**Comment by nicobao** [comment]
I recommend https://github.com/ggml-org/llama.vim

**Comment by ya-hitho** [comment]
I've been using supermaven. I feel it is way faster than copilot and give more precise suggestions.

**Reply by ns-bsena** [reply]
Unfortunately died

**Comment by lqz-hub** [comment]
I recommend  https://github.com/Kurama622/llm.nvim, which just dependens on the ollama and curl
